cmake/BuildBenchmark.cmake

```diff
@@ -0,0 +1,61 @@
+set(BENCHMARK_VERSION v1.9.0)
+set(BENCHMARK_HASH 21a2604efeded8b4cbabc72f3e1c7a2a)
+set(BENCHMARK_INSTALL_DIR ${BUNDLED_LIBS_INSTALL_DIR}/benchmark-prefix)
+set(BENCHMARK_INCLUDE_DIR ${BENCHMARK_INSTALL_DIR}/dest/include)
+
+set(BENCHMARK_LIB ${BENCHMARK_INSTALL_DIR}/build/src/libbenchmark.a)
+set(BENCHMARK_LIB_MAIN ${BENCHMARK_INSTALL_DIR}/build/src/libbenchmark_main.a)
+
+list(APPEND BENCHMARK_CMAKE_FLAGS
```

Comment:
`perf/` contains executables that defined outside tarantool build system, for example `small.perftest`.

CMake configuration is failed with command `cmake -S . -B build -DENABLE_BUNDLED_LIBBENCHMARK=ON -DCMAKE_BUILD_TYPE=Release` due to `return()` (https://github.com/tarantool/small/blob/d118d3577233e6769645e8bc273df2c66da91d8f/perf/CMakeLists.txt#L11):

```
-- Configuring done (0.3s)  
CMake Error at perf/CMakeLists.txt:59 (add_custom_command):
  Error evaluating generator expression:           

    $<TARGET_FILE:small.perftest>
                                                                         
  No target "small.perftest"                                             
Call Stack (most recent call first): 
  perf/CMakeLists.txt:102 (create_perf_test_target)

                                    
CMake Error at perf/CMakeLists.txt:61 (add_custom_target):
  Error evaluating generator expression:           

    $<TARGET_FILE:small.perftest>
                                    
  No target "small.perftest"                                                                                                                       
Call Stack (most recent call first): 
  perf/CMakeLists.txt:102 (create_perf_test_target)


CMake Error at perf/CMakeLists.txt:59 (add_custom_command):
  Error evaluating generator expression:

    $<TARGET_FILE:small.perftest>

  No target "small.perftest"
```

To fix this we need to propagate `BENCHMARK_LIBRARIES` to `small` library:

- introduce `benchmark_FOUND` in `cmake/BuildBenchmark.cmake`
- move a condition for `ENABLE_BUNDLED_BENCHMARK` and with definition of `BENCHMARK_LIBRARIES` from `perf/CMakeLists.txt` to `CMakeLists.txt` and put it *before* `add_subdirectory(src)` (it is needed for `small`)
- apply a patch to `src/lib/small/perf/CMakeLists.txt` in a small repo and then bump small version in tarantool repository

Full patchset in Tarantool repository:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index a8a3990317..3c8cbedf9b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -772,14 +772,37 @@ include (cmake/package.cmake)
 #
 include (cmake/rpm.cmake)
 
-add_subdirectory(src)
 add_subdirectory(extra)
 add_subdirectory(test)
 if (CMAKE_BUILD_TYPE STREQUAL "Debug")
     message(AUTHOR_WARNING "Benchmarks are available only in release build")
 else ()
+
+    if(NOT ENABLE_BUNDLED_LIBBENCHMARK)
+        find_package(benchmark QUIET)
+        if(NOT ${benchmark_FOUND})
+            message(AUTHOR_WARNING "Google Benchmark library was not found")
+            set(MSG "Target test-c-perf is dummy, Google Benchmark library was not found")
+            add_custom_target(test-c-perf
+                              COMMAND ${CMAKE_COMMAND} -E cmake_echo_color --red ${MSG}
+                              COMMENT ${MSG}
+            )
+            add_custom_target(test-perf
+                              DEPENDS test-c-perf test-lua-perf
+                             COMMENT "Running performance tests"
+            )
+            return()
+        endif()
+        set(BENCHMARK_LIBRARIES benchmark::benchmark)
+    else()
+        include(BuildBenchmark)
+        add_dependencies(build_bundled_libs ${BENCHMARK_LIBRARIES})
+        include_directories(${BENCHMARK_INCLUDE_DIRS})
+    endif()
+
     add_subdirectory(perf)
 endif ()
+add_subdirectory(src)
 add_subdirectory(doc)
 
 option(WITH_NOTIFY_SOCKET "Enable notifications on NOTIFY_SOCKET" ON)
diff --git a/cmake/BuildBenchmark.cmake b/cmake/BuildBenchmark.cmake
index d68441c89c..2884966dda 100644
--- a/cmake/BuildBenchmark.cmake
+++ b/cmake/BuildBenchmark.cmake
@@ -51,6 +51,7 @@ add_custom_target(bundled-libbenchmark
 set(BENCHMARK_LIBRARIES bundled-benchmark-main bundled-benchmark)
 
 set(BENCHMARK_INCLUDE_DIRS ${BENCHMARK_INCLUDE_DIR})
+set(benchmark_FOUND TRUE)
 
 unset(BENCHMARK_VERSION)
 unset(BENCHMARK_HASH)
diff --git a/perf/CMakeLists.txt b/perf/CMakeLists.txt
index 82f3a8df62..a2dd2b29a7 100644
--- a/perf/CMakeLists.txt
+++ b/perf/CMakeLists.txt
@@ -15,27 +15,6 @@ separate_arguments(BENCH_CMD_SEPARATE UNIX_COMMAND ${BENCH_CMD})
 
 add_subdirectory(lua)
 
-if(NOT ENABLE_BUNDLED_LIBBENCHMARK)
-    find_package(benchmark QUIET)
-    if(NOT ${benchmark_FOUND})
-        message(AUTHOR_WARNING "Google Benchmark library was not found")
-        set(MSG "Target test-c-perf is dummy, Google Benchmark library was not found")
-        add_custom_target(test-c-perf
-                          COMMAND ${CMAKE_COMMAND} -E cmake_echo_color --red ${MSG}
-                          COMMENT ${MSG}
-        )
-        add_custom_target(test-perf
-                          DEPENDS test-c-perf test-lua-perf
-                          COMMENT "Running performance tests"
-        )
-        return()
-    endif()
-    set(BENCHMARK_LIBRARIES benchmark::benchmark)
-else()
-    include(BuildBenchmark)
-    add_dependencies(build_bundled_libs bundled-libbenchmark)
-    include_directories(${BENCHMARK_INCLUDE_DIRS})
-endif()
 
 include_directories(${MSGPUCK_INCLUDE_DIRS})
 include_directories(${PROJECT_SOURCE_DIR}/src/box)
```

Patch in small repository:

```diff
diff --git a/perf/CMakeLists.txt b/perf/CMakeLists.txt
index de1ddc5..2596210 100644
--- a/perf/CMakeLists.txt
+++ b/perf/CMakeLists.txt
@@ -5,13 +5,23 @@ if (CMAKE_BUILD_TYPE STREQUAL "Debug")
     return()
 endif()
 
-find_package(benchmark QUIET)
+# Handle a case when Benchmark variables (BENCHMARK_LIBRARIES,
+# benchmark_FOUND) are defined outside, like Tarantool, that uses
+# a bundled Benchmark library.
 if (NOT ${benchmark_FOUND})
-    message(AUTHOR_WARNING "Google Benchmark submodule not found")
-    return()
+    find_package(benchmark QUIET)
+    if (NOT ${benchmark_FOUND})
+        message(AUTHOR_WARNING "Google Benchmark submodule not found")
+        return()
+    endif()
+    set(BENCHMARK_LIBRARIES benchmark::benchmark)
 endif()
 
 include_directories("${PROJECT_SOURCE_DIR}")
 
+# if (NOT ${BENCHMARK_LIBRARIES})
+#     set(BENCHMARK_LIBRARIES benchmark::benchmark)
+# endif()
+
 add_executable(small.perftest small.cc)
-target_link_libraries(small.perftest small benchmark::benchmark)
+target_link_libraries(small.perftest small ${BENCHMARK_LIBRARIES})
```

Reply:
I've created the corresponding PR (tarantool/small#100) for these changes.
This branch is updated with a temporary commit to test the behaviour.